doc/config.rst
.. _config:

Configuration values
====================

``chirpy`` can be configured in several ways:

- Most values have sensible defaults where possible
- :ref:`Scripts <scripts>` take command-line options that can set some
  configuration values
- The values can be set programmatically by accessing variables in the
  ``chirpy.config`` module
- Some values can also be set from a program's environment by setting
  corresponding environment variables

The use of environment variables is recommended. This allows all tools
in a pipeline to easily share a configuration so that, for example,
they will all use the same MQTT broker. The other mechanisms are
available for more complicated cases.

Node identification
-------------------

+--------------------+------------------+----------------------------------------+--------------------------+
| Variable           | Default          | Description                            | Environment variable     |
+====================+==================+========================================+==========================+
| ``nodeIdentifier`` | "chirpy"         | Node identification string, needed for | ``CHIRPY_NODE``          |
|                    |                  | systems with several sensors           |                          |
+--------------------+------------------+----------------------------------------+--------------------------+
| ``nodeLatitude``   | 56.33404         | Node latitude                          | ``CHIRPY_NODE_LAT``      |
+--------------------+------------------+----------------------------------------+--------------------------+
| ``nodeLongitude``  | -2.79955         | Node longitude                         | ``CHIRPY_NODE_LON``      |
+--------------------+------------------+----------------------------------------+--------------------------+

The node identifier will be passed in all JSON objects within
``chirpy``, allowing the source of observations to be determined.

The latitude and longitude are used when computing sunrise and sunset
times (for avoiding sampling at night), and defaults to `St Andrews`_.
There's no need to be too precise about these values: within a degree
will be enough to enable this functionality.

However, for systems with lots of sensors spread over an area, or
where the exact locations of the observations are important, providing
a more precise location may be useful.

.. _mqtt-config:

MQTT broker
-----------

These values are used only when reporting observations *via* an MQTT broker.

+-------------------+------------------+---------------------------------------+--------------------------+
| Variable          | Default          | Description                           | Environment variable     |
+===================+==================+=======================================+==========================+
| ``mqttHost``      | None             | MQTT broker hostname or IP address    | ``CHIRPY_MQTT_HOST``     |
+-------------------+------------------+---------------------------------------+--------------------------+
| ``mqttUsername``  | None             | MQTT username                         | ``CHIRPY_MQTT_USERNAME`` |
+-------------------+------------------+---------------------------------------+--------------------------+
| ``mqttPassword``  | None             | MQTT password                         | ``CHIRPY_MQTT_PASSWORD`` |
+-------------------+------------------+---------------------------------------+--------------------------+

.. _mesh-config:

Meshtastic mesh network
-----------------------

These values are used only when reporting observations *via* a
Meshtastic mesh network. The Meshtastic device used as gateway can be
identified either by a hostname or by a local serial device.

+-----------------+-------------+------------------------------------------+-------------------------+
| Variable        | Default     | Description                              |  Environment variable   |
+=================+=============+==========================================+=========================+
| ``meshHost``    | None        | Meshtastic device hostname or IP address |  ``CHIRPY_MESH_HOST``   |
+-----------------+-------------+------------------------------------------+-------------------------+
| ``meshPort``    | None        | Meshtastic device serial port            | ``CHIRPY_MESH_PORT``    |
+-----------------+-------------+------------------------------------------+-------------------------+
| ``meshChannel`` | None        | Meshtastic channel                       | ``CHIRPY_MESH_CHANNEL`` |
+-----------------+-------------+------------------------------------------+-------------------------+

SQLite database
---------------

These variables set up the SQLite database used for storing observations.

+-------------------+------------------+---------------------------------------+----------------------+
| Variable          | Default          | Description                           | Environment variable |
+===================+==================+=======================================+======================+
| ``sqlitedb``      | "chirpy.db"      | File name for SQLite database         | ``CHIRPY_SQLITE_DB`` |
+-------------------+------------------+---------------------------------------+----------------------+

System tuning
-------------

These tuning values all have sensible defaults.

+-------------------------+--------------------+----------------------------------------+----------------------+
| Variable                | Default            | Description                            | Environment variable |
+=========================+====================+========================================+======================+
| ``logLevel``            | ``logging.INFO``   | Log level                              | ``CHIRPY_LOG_LEVEL`` |
+-------------------------+--------------------+----------------------------------------+----------------------+
| ``sampleDuration``      | 5                  | Length of each acoustic sample (s)     |                      |
+-------------------------+--------------------+----------------------------------------+----------------------+
| ``confidenceThreshold`` | 0.2                | Minimal confidence for an              |                      |
|                         |                    | observation to be reported             |                      |
+-------------------------+--------------------+----------------------------------------+----------------------+
| ``nighttimeOffset``     | 60                 | Time from sunrise/sunset considered    |                      |
|                         |                    | to be night-time (min)                 |                      |
+-------------------------+--------------------+----------------------------------------+----------------------+
| ``heartbeat``           | 5 * 60             | Interval between heartbeats (s)        |                      |
+-------------------------+--------------------+----------------------------------------+----------------------+

The ``CHIRPY_LOG_LEVEL`` value can be a number or an expression. It is
evaluated in an environment where the ``logging`` module is imported,
so a value such as "logging.CRITICAL" will set the logging level to
``logging.CRITICAL``.



.. _St Andrews: https://en.wikipedia.org/wiki/St_Andrews


chirpy/config.py
# Configuration variables
#
# Copyright (C) 2025--2026 Simon Dobson
#
# This is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This software is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this software. If not, see <http://www.gnu.org/licenses/gpl.html>.

# ---------- Node identifier ----------

nodeIdentifier = "chirpy"
nodeLatitude = 56.33404
nodeLongitude = -2.79955


# ---------- MQTT broker ----------

mqttHost = None
mqttUsername = None
mqttPassword = None


# ---------- Meshtastic ----------

meshHost = "localhost"
meshPort = None


# ---------- SQLite database ----------

sqlitedb = "chirpy.db"


# ---------- System tuning ----------

import logging

logLevel = logging.INFO
sampleDuration = 5
confidenceThreshold = 0.2
nighttimeOffset = 60
heartbeat = 5 * 60
